src/GridLineEstimator.cpp

```diff
@@ -107,6 +104,41 @@ GridLineEstimator::GridLineEstimator(
                                                                        10);
 
     yaw_pub_ = local_nh.advertise<iarc7_msgs::Float64Stamped>("line_yaw", 10);
+
+    // Create the canny edge detector with dummy settings that will be replaced
+    // on the first called to onSettingsChanged
+    gpu_canny_edge_detector_ = cv::cuda::createCannyEdgeDetector(
```

Comment:
The only functions that can load the parameters have to be called by dynamic reconfigure currently. For dynamic reconfigure to be setup it has to be passed a lambda with the vision algorithms so that it can the updateSettings methods in those objects. That requires the vision algorithm objects to be created before dynamic reconfigure is started.

IIRC there is no guarantee that dynamic reconfigure will call the callback immediately so we might have a race condition here too.

Reply:
The parameters are loaded by functions `get...Settings` in `VisionNode.cpp` before dynamic reconfigure is set up

Reply:
getDynamicSettings is not and that one loads all the settings needed for these objects.

Reply:
Wow I totally missed that

Reply:
You're right that there's still a race condition.  Let's call `getDynamicSettings` before we call those constructors, with a dummy config and `false` for the `ran` argument, I think that fixes it?

Reply:
I'm going to break the param server side of getDynamicSettings into getOpticalFlowSettings and getLineEstimatorSettings. I'll call those two functions in the beginning. Then getDynamicSettings can continued to behave as before and we don't have to make a blank config object or do the ugly ran argument hack.

Reply:
:+1:

Reply:
Fixed